File: libs/core/src/declarations/merged_namespace_declaration.cpp
/**
 * @file merged_namespace_declaration.cpp
 * @author Francesco Corso
 * @date 8/13/2026
 * @brief
 */
module;

#include <libassert/assert-macros.hpp>

module prism.core:declarations.merged_namespace_declaration.impl;

import :declarations.merged_namespace_declaration;
import :memory.buffer_pool;
import :declarations.visit;
import :util.overload;

namespace prism
{
    namespace
    {
        void add_namespace_to_children(const PooledVector<RefCountPtr<const SingleNamespaceDeclaration>> &namespaces,
                                       const bool all_namespaces_has_same_name,
                                       PooledVector<RefCountPtr<const MergedDeclaration>> &children)
        {
            if (namespaces.empty())
                return;

            if (all_namespaces_has_same_name)
            {

                children.emplace_back(MergedNamespaceDeclaration::create(ImmutableArray{std::from_range, namespaces}));
            }
            else
            {
                PooledMap<Name, PooledVector<RefCountPtr<const SingleNamespaceDeclaration>>> namespaces_by_name;
                for (auto &n : namespaces)
                {
                    namespaces_by_name[n->name()].push_back(n);
                }

                for (const auto &grouped : namespaces_by_name | std::views::values)
                {
                    children.emplace_back(MergedNamespaceDeclaration::create(ImmutableArray{std::from_range, grouped}));
                }
            }
        }
    } // namespace

    MergedNamespaceDeclaration::MergedNamespaceDeclaration(ConstructTag, SingleNamespaceList declarations)
        : MergedDeclaration{!declarations.empty() ? declarations.front()->name() : KnownName::none,
                            DeclarationKind::namespace_},
          declarations_{std::move(declarations)}
    {
    }

    RefCountPtr<const MergedNamespaceDeclaration> MergedNamespaceDeclaration::create(SingleNamespaceList declarations)
    {
        return make_ref_counted<const MergedNamespaceDeclaration>(construct_tag, std::move(declarations));
    }

    RefCountPtr<const MergedNamespaceDeclaration> MergedNamespaceDeclaration::create(
        const SingleNamespaceDeclaration &declaration)
    {
        return create(ImmutableArray{declaration.shared_from_this()});
    }

    LexicalSortKey MergedNamespaceDeclaration::get_lexical_sort_key(const Compilation &compilation) const
    {
        LexicalSortKey sort_key{declarations_[0]->name_location(), compilation};
        for (const auto declaration : declarations_ | std::views::drop(1))
        {
            sort_key = LexicalSortKey::first(sort_key, LexicalSortKey{declaration->name_location(), compilation});
        }

        return sort_key;
    }

    ImmutableArray<Location> MergedNamespaceDeclaration::name_locations() const
    {
        return declarations_ |
               std::views::transform([](const RefCountPtr<const SingleNamespaceDeclaration> &declaration) -> auto &
                                     { return declaration->name_location(); }) |
               std::ranges::to<ImmutableArray<Location>>();
    }

    const ImmutableArray<RefCountPtr<const MergedDeclaration>> &MergedNamespaceDeclaration::members() const
    {
        return members_.get_or_compute([this] { return make_children(); });
    }

    const ImmutableHashSet<Name> &MergedNamespaceDeclaration::member_names() const
    {
        return member_names_.get_or_compute(
            [this]
            {
                return declarations_ |
                       std::views::transform(
                           [](const RefCountPtr<const SingleNamespaceDeclaration> &declaration) -> auto &
                           { return declaration->member_names(); }) |
                       std::views::join | std::ranges::to<ImmutableHashSet>();
            });
    }

    ImmutableArray<RefCountPtr<const MergedDeclaration>> MergedNamespaceDeclaration::make_children() const
    {
        PooledVector<RefCountPtr<const SingleNamespaceDeclaration>> namespaces;
        bool all_namespaces_has_same_name = true;

        for (const auto &child :
             declarations_ |
                 std::views::transform([](const RefCountPtr<const SingleNamespaceDeclaration> &declaration) -> auto &
                                       { return declaration->members(); }) |
                 std::views::join)
        {
            visit(*child,
                  Overload{[&](const SingleNamespaceDeclaration &as_namespace)
                           {
                               if (!namespaces.empty() && all_namespaces_has_same_name &&
                                   as_namespace.name() != namespaces.front()->name())
                               {
                                   all_namespaces_has_same_name = false;
                               }

                               namespaces.emplace_back(as_namespace.shared_from_this());
                           }});
        }

        PooledVector<RefCountPtr<const MergedDeclaration>> children;

        add_namespace_to_children(namespaces, all_namespaces_has_same_name, children);

        return ImmutableArray{std::from_range, children};
    }
} // namespace prism

File: src/Prism.Core/Declarations/MergedNamespaceDeclaration.cs
using System.Collections.Immutable;
using Prism.Core.Compiling;
using Prism.Core.Diagnostics;
using Prism.Core.Symbols;
using ZLinq;

namespace Prism.Core.Declarations;

internal sealed class MergedNamespaceDeclaration : MergedDeclaration
{
    private MergedNamespaceDeclaration(ImmutableArray<SingleNamespaceDeclaration> declarations)
        : base(!declarations.IsEmpty ? declarations[0].Name : "")
    {
        Declarations = declarations;
    }

    public static MergedNamespaceDeclaration Create(
        ImmutableArray<SingleNamespaceDeclaration> declarations
    )
    {
        return new MergedNamespaceDeclaration(declarations);
    }

    public override DeclarationKind Kind => DeclarationKind.Namespace;

    public LexicalSortKey GetLexicalSortKey(Compilation compilation)
    {
        var sortKey = new LexicalSortKey(Declarations[0].NameLocation, compilation);
        foreach (var declaration in Declarations.AsValueEnumerable().Skip(1))
        {
            sortKey = LexicalSortKey.First(
                sortKey,
                new LexicalSortKey(declaration.NameLocation, compilation)
            );
        }
        return sortKey;
    }

    public ImmutableArray<SingleNamespaceDeclaration> Declarations { get; }

    public ImmutableArray<Location> NameLocations
    {
        get { return [.. Declarations.Select(Location (d) => d.NameLocation)]; }
    }

    public new ImmutableArray<MergedDeclaration> Members
    {
        get
        {
            if (!field.IsDefault)
            {
                return field;
            }

            ImmutableInterlocked.InterlockedCompareExchange(ref field, MakeChildren(), default);
            return field;
        }
    }

    public ImmutableHashSet<string> MemberNames
    {
        get
        {
            if (field is not null)
            {
                return field;
            }

            Interlocked.CompareExchange(
                ref field,
                [.. Declarations.AsValueEnumerable().SelectMany(d => d.MemberNames)],
                null
            );
            return field;
        }
    }

    protected override ImmutableArray<MergedDeclaration> GetMergedMembers()
    {
        return Members;
    }

    private ImmutableArray<MergedDeclaration> MakeChildren()
    {
        var namespaces = ImmutableArray.CreateBuilder<SingleNamespaceDeclaration>();
        var allNamespacesHaveSameName = true;

        var types = ImmutableArray.CreateBuilder<SingleTypeDeclaration>();
        var allTypesHaveSameIdentity = true;

        foreach (
            var child in Declarations
                .AsValueEnumerable()
                .SelectMany(d => d.Members.AsValueEnumerable())
        )
        {
            switch (child)
            {
                case SingleNamespaceDeclaration asNamespace:
                    if (
                        namespaces.Count > 0
                        && allNamespacesHaveSameName
                        && asNamespace.Name != namespaces[0].Name
                    )
                    {
                        allNamespacesHaveSameName = false;
                    }

                    namespaces.Add(asNamespace);
                    break;
                case SingleTypeDeclaration asType:
                    if (
                        types.Count > 0
                        && allTypesHaveSameIdentity
                        && asType.Identity != types[0].Identity
                    )
                    {
                        allTypesHaveSameIdentity = false;
                    }

                    types.Add(asType);
                    break;
            }
        }

        var children = ImmutableArray.CreateBuilder<MergedDeclaration>();
        AddNamespacesToChildren(namespaces, allNamespacesHaveSameName, children);
        AddTypesToChildren(types, allTypesHaveSameIdentity, children);

        return children.DrainToImmutable();
    }

    private static void AddNamespacesToChildren(
        ImmutableArray<SingleNamespaceDeclaration>.Builder namespaces,
        bool allNamespacesHaveSameName,
        ImmutableArray<MergedDeclaration>.Builder children
    )
    {
        if (namespaces.Count == 0)
            return;

        if (allNamespacesHaveSameName)
        {
            children.Add(Create(namespaces.DrainToImmutable()));
            return;
        }

        foreach (var grouping in namespaces.GroupBy(ns => ns.Name))
        {
            children.Add(Create([.. grouping]));
        }
    }

    private static void AddTypesToChildren(
        ImmutableArray<SingleTypeDeclaration>.Builder types,
        bool allTypesHaveSameName,
        ImmutableArray<MergedDeclaration>.Builder children
    )
    {
        if (types.Count == 0)
            return;

        if (allTypesHaveSameName)
        {
            children.Add(MergedTypeDeclaration.Create(types.DrainToImmutable()));
            return;
        }

        foreach (var grouping in types.GroupBy(ns => ns.Name))
        {
            children.Add(MergedTypeDeclaration.Create([.. grouping]));
        }
    }
}
